src/__rt/mod.rs

```diff
@@ -43,14 +43,26 @@ pub enum Mu {}
 
 thread_local!(static CTX: Cell<*mut task::Context<'static>> = Cell::new(ptr::null_mut()));
 
-struct Reset<'a>(*mut task::Context<'static>, &'a Cell<*mut task::Context<'static>>);
+pub struct Reset<'a>(*mut task::Context<'static>, &'a Cell<*mut task::Context<'static>>);
+
+impl<'a> Reset<'a> {
+    pub fn ctx(&mut self) -> &mut task::Context {
+        if self.0 == ptr::null_mut() {
+            panic!("Cannot use `await!` outside of an `async` function.")
+        }
+        unsafe { mem::transmute(self.0) }
+    }
+}
 
 impl<'a> Drop for Reset<'a> {
     fn drop(&mut self) {
         self.1.set(self.0);
     }
 }
 
-pub fn get_ctx() -> *mut task::Context<'static> {
-    CTX.with(|cell| cell.get())
+pub fn in_ctx<F: FnOnce(Reset) -> T, T>(f: F) -> T {
```

Comment:
In order for this to be safe, `Reset` can't be given away to the inner function, because it could `mem::forget` it. Changing `F` to be `FnOnce(&mut Context)` should work.